src/TauBounds.cpp

```diff
@@ -5,62 +5,127 @@
+  int mig, target, source;
+  double age;
+  for (int gen = 0; gen < dataSetup.numLoci; gen++) {
+    for (int i = 0; i < genetree_migs[gen].num_migs; i++) {
+      mig = genetree_migs[gen].living_mignodes[i];
+      target = genetree_migs[gen].mignodes[mig].target_pop;
+      source = genetree_migs[gen].mignodes[mig].source_pop;
+      age = genetree_migs[gen].mignodes[mig].age;
+      propagateUpperBoundDownwards(target, age);
+      propagateUpperBoundDownwards(source, age);
+    }
+  }
+}
+
+void propagateUpperBoundDownwards(int pop, double bound) {
+  tau_ubounds[pop] = fmin(tau_ubounds[pop], bound);
+  for (int dec = 0; dec < dataSetup.popTree->numPops; dec++) {
+    if (isAncestralTo(pop, dec)) {
+      tau_ubounds[dec] = fmin(tau_ubounds[dec], bound);
+    }
+  }
 }
 
-int calculateLociTauBounds(int nodeId, int gen) {
+
+int calculateLocusTauBounds(int nodeId, int gen) {
   LikelihoodNode *currentNode = getNode(nodeId, gen);
   int pop = getNodePop(nodeId, gen);
 
   if (isLeafPop(pop))
-    return pop;
+    return migLcaPop(nodeId, gen, pop);
 
-  int leftLca = calculateLociTauBounds(currentNode->leftSon, gen);
-  int rightLca = calculateLociTauBounds(currentNode->rightSon, gen);
+  int leftLca = calculateLocusTauBounds(currentNode->leftSon, gen);
+  int rightLca = calculateLocusTauBounds(currentNode->rightSon, gen);
 
   int lca_pop = lca_pops[leftLca][rightLca];
+  tau_ubounds[lca_pop] = fmin(tau_ubounds[lca_pop], currentNode->age);
+  return migLcaPop(nodeId, gen, lca_pop);
+}
 
-  tau_bounds[lca_pop] = fmin(tau_bounds[lca_pop], currentNode->age);
-
-  return lca_pop;
+/**
+ *  if nodeId is right below a migration event, returns the source population of the migration event.
+ *  Otherwise, returns defaultPop
+ */
+int migLcaPop(int nodeId, int gen, int defaultLcaPop) {
+  for (int i = 0; i < genetree_migs[gen].num_migs; i++) {
+    int mig = genetree_migs[gen].living_mignodes[i];
+    if (genetree_migs[gen].mignodes[mig].gtree_branch == nodeId)
+      return genetree_migs[gen].mignodes[mig].source_pop;
+  }
+  return defaultLcaPop;
 }
 
 void propagateBoundsDownPopTree() {
   for (int pop = 0; pop < dataSetup.popTree->numPops; pop++) {
-    updateTauBoundsOfDescendants(pop, tau_bounds[pop]);
+    updateTauBoundsOfDescendants(pop, tau_ubounds[pop]);
   }
 }
 
 void updateTauBoundsOfDescendants(int pop, double bound) {
   if (isLeafPop(pop)) return;
 
-  tau_bounds[pop] = fmin(tau_bounds[pop], bound);
+  tau_ubounds[pop] = fmin(tau_ubounds[pop], bound);
 
   updateTauBoundsOfDescendants(getSon(pop, LEFT), bound);
   updateTauBoundsOfDescendants(getSon(pop, RIGHT), bound);
 }
 
-
 void printTauBoundsHeader(FILE *file) {
   fprintf(file, "iteration");
   for (int pop = 0; pop < dataSetup.popTree->numPops; pop++) {
     if (isLeafPop(pop)) continue;
-    fprintf(file, "\t%s_bound\t%s_tau", getPopName(pop), getPopName(pop));
+    fprintf(file, "\t%s_ubound\t%s_lbound\t%s_tau", getPopName(pop), getPopName(pop), getPopName(pop));
```

Comment:
the `tau_bounds.tsv` output file now contains another column per comb. btw, @gphocs-dev , do we also need to calculate tau bounds for clades? let's discuss!

Reply:
I'm pretty sure that the same bounds apply for each tau in every context of a clade reference model. Combs might be slightly different because tau_min acts as another lower bound.